src/graph_visualization.cpp

```diff
@@ -0,0 +1,463 @@
+    nodes.push_back(node);
+    nodes[i].id = i;
+    for (int j = 0; j < input[i].size(); j++) {
+      nodes[i].succ.push_back(input[i][j]);
+    }
+  }
+}
+
+void predReinit(std::vector<TreeNode> &nodes) {
+  for (int i = 0; i < nodes.size(); i++) {
+    nodes[i].pred = {};
+  }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      for (int k = 0; k < nodes.size(); k++) {
+        if (nodes[k].id == nodes[i].succ[j]) {
+          nodes[k].pred.push_back(nodes[i].id);
+        }
+      }
+    }
+  }
+}
+
+void degReinit(std::vector<TreeNode> &nodes) {
+  for (int i = 0; i < nodes.size(); i++) {
+    nodes[i].degP = 0;
+    nodes[i].degM = 0;
+  }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      nodes[i].degP++;
+
+      for (int k = 0; k < nodes.size(); k++) {
+        if (nodes[k].id == nodes[i].succ[j]) {
+          nodes[k].degM++;
+        }
+      }
+    }
+  }
+}
+
+void algorithmDFS(int &exit, std::vector<TreeNode> &nodes, std::vector<bool> &usedId, TreeNode::nodeId id, TreeNode::nodeId prevId) {
+  if (exit == 0) {
+    usedId[id] = true;
+
+    int numb = -1;
+    for (int i = 0; i < nodes.size(); i++) {
+      if (nodes[i].id == id) {
+        numb = i;
+        break;
+      }
+    }
+
+    for (int i = 0; i < nodes[numb].succ.size(); i++) {
+      if (!usedId[nodes[numb].succ[i]]) {
+        algorithmDFS(exit, nodes, usedId, nodes[numb].succ[i], id);
+      } else {
+        if (id != prevId) {
+          printf("Graph has cycle. \n");
+          exit = 1;
+          break;
+        }
+      }
+    }
+  }
+}
+
+void cycleExistsDFS(std::vector<TreeNode> &nodes) {
+  std::vector<bool> usedId = {};
+  for (int i = 0; i < nodes.size(); i++) {
+    usedId.push_back(false);
+  }
+  int exit = 0;
+  for (int i = 0; i < nodes.size(); i++) {
+    if (exit == 1) {
+      break;
+    }
+    if (!usedId[nodes[i].id]) {
+      algorithmDFS(exit, nodes, usedId, nodes[i].id, -1);
+    }
+  }
+}
+
+void heapify(std::vector<TreeNode> &nodes, int n, int i) {
+  int largest = i;
+  int l = 2 * i + 1;
+  int r = 2 * i + 2;
+
+  if (l < n && nodes[l].id > nodes[largest].id) {
+    largest = l;
+  }
+
+  if (r < n && nodes[r].id > nodes[largest].id) {
+    largest = r;
+  }
+
+  if (largest != i) {
+    std::swap(nodes[i], nodes[largest]);
+    heapify(nodes, n, largest);
+  }
+}
+
+void idHeapSort(std::vector<TreeNode> &nodes) {
+  int n = 0;
+  for (int i = 0; i < nodes.size(); i++) {
+    n++;
+  }
+
+  for (int i = n / 2 - 1; i >= 0; i--) {
+    heapify(nodes, n, i);
+  }
+
+  for (int i = n - 1; i >= 0; i--) {
+    std::swap(nodes[0], nodes[i]);
+    heapify(nodes, i, 0);
+  }
+}
+
+TreeNode getTreeNodeById(std::vector<TreeNode> &nodes, TreeNode::nodeId id) {
+  TreeNode node = {};
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].id == id) {
+      node = nodes[i];
+      break;
+    }
+  }
+  return node;
+}
```

Comment:
Why return a copy? Return a pointer with `nullptr` indicating `not-found` or return a reference and `assert`.
```suggestion
TreeNode *getTreeNodeById(std::vector<TreeNode> &nodes, TreeNode::nodeId id) {
    if (id >= 0 && static_cast<size_t>(id) < node.size()) {
      return nodes.data() + id;
    }
    return nullptr;
}
```